### PR TITLE
version 2.0

### LINKED ISSUE #46: Updating public-repo deletes all previously uploaded apps
This is a really silly design issue...

When you update an app, the `/db/app/<app-dir>` is deleted before the new version is installed. If the app stores its data inside its app dir then all its data is deleted.

One example of this, is this app, i.e. the `public-repo`. Updating the public repo causes all the apps on the public repo to be lost. This is quite ridiculous!

### PR DESCRIPTION
**Pull request by joewiz**
Closes #46 and #49.

BREAKING CHANGES

- Packages are now stored outside the `public-repo` collection, in a sibling `public-repo-data` collection. Packages are no longer deleted when upgrading the public repo app.
- To prevent permissions issues, uploading and publishing apps is limited to members of the `repo` group.
- The `public-repo-data` collection is split into 3 sub-collections: `packages`, `icons`, and `metadata`. 
- Despite the changes in storage, no public API changes have been introduced: the HTML listings, download URLs, API endpoints, etc. are unchanged.
- The metadata files have been renamed from `packages.xml` (introduced in #52) to `raw-packages.xml` (with the structure `/raw-packages/package`) and from `apps.xml` to `package-groups.xml` (with the structure `/package-groups/package-group`). 
- The app now requires eXist 5. 
- The app's dependency on the crypto library has been removed.
- Due to the breaking changes, I've gone ahead and incremented the version number in this PR to 2.0.0. 

To aid in upgrading existing installations to this location and format, the PR includes a script that performs the upgrade from the v1 format to the v2 format—which must be done BEFORE upgrading the app itself. I'll send out instructions on how to use this script to exist-open and Slack when the PR is merged.

I've tested the PR extensively with eXist 5.2.0, but DO NOT MERGE BEFORE SOMEONE WHO HAS TESTED IT APPROVES. (~~The app is broken in 5.3.0-SNAPSHOT due to its dependency on the EXPath Crypto library, which is broken.~~ The app is compatible with 5.3.0-SNAPSHOT since it no longer depends on the Crypto library.)

To test:

1. install public-repo v1.0.3 or v1.1.0 on an eXist 5.1.1 system (the version running on the exist-db.org server now) or newer.
2. upload a representative sample of xars and publish them (from admin.html, which triggers an AJAX request served by `upload.xq`); filenames of uploaded xars should appear in the "Uploaded Files" section 
3. run the upgrade script in `modules/upgrade-to-v2-storage.xq`
4. install this PR
5. the xar listing on the landing page, package-specific listings on interior pages, and all endpoints should respond like normal
    - landing page xar listing: http://localhost:8080/exist/apps/public-repo
    - package-specific listings, with version parameters: http://localhost:8080/exist/apps/public-repo/packages/semver-xq.html?eXist-db-min-version=4.7.0
    - xar download service, served by `controller.xql` which forwards requests to the data collection: http://localhost:8080/exist/apps/public-repo/public/eXide-2.4.10.xar
    - icon download service, served by `controller.xql`, which forwards requests to the data collection: http://localhost:8080/exist/apps/public-repo/public/eXide-2.4.10.png
    - atom feed, served by `modules/feed.xq`: http://localhost:8080/exist/apps/public-repo/feed.xml
    - version-specific listing needed for packageservice, served by list.xq: http://localhost:8080/exist/apps/public-repo/public/apps.xml?version=5.2.0 (and lower versions to confirm correct filtering)
    - package retrieval service needed by eXist build tools and the `repo:install-and-deploy` function, served by `modules/find.xq`: http://localhost:8080/exist/apps/public-repo/find?abbrev=eXide&processor=5.2.0 (append "&info=yes" for a debug version)
    - .zip file extension download service, served by `modules/download-xar-zip.xq`: http://localhost:8080/exist/apps/public-repo/public/eXide-2.4.10.xar.zip
    - rebuild package metadata service, served by update.xq
6. upload new xars to confirm the can be published.
7. check the various pages and endpoints again
6. for good measure, reinstall the PR and confirm that xars are not deleted.

Thanks to @line-o for making public-repo scalable again with #52. This was the breakthrough that inspired me to dig deeper into this problem, at long last.

### CODE REVIEW

File: controller.xql [hunk 1]
```diff
@@ -91,6 +92,16 @@ else if (ends-with($exist:resource, ".html")) then
         </view>
     </dispatch>
 
+else if (contains($exist:path, "/public/") and ends-with($exist:resource, ".xar")) then
+    <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
+        <forward url="../../public-repo-data/packages/{$exist:resource}"/>
+    </dispatch>
+
+else if (contains($exist:path, "/public/") and (ends-with($exist:resource, ".png") or ends-with($exist:resource, ".svg"))) then
```
Comment (joewiz) [inline controller.xql:100]: TODO check if we support other icon image formats

File: repo.xml [hunk 2]
```diff
@@ -8,10 +8,17 @@
...
     <target>public-repo</target>
-    <prepare>pre-install.xql</prepare>
-    <finish>post-install.xql</finish>
+    <prepare>pre-install.xq</prepare>
+    <finish>post-install.xq</finish>
     <permissions xmlns:repo="http://exist-db.org/xquery/repo" password="repo" user="repo" group="repo" mode="rw-rw-r--"/>
     <changelog>
+        <change version="2.0.0">
+            <ul xmlns="http://www.w3.org/1999/xhtml">
+                <li>IMPORTANT: You must first run the upgrade script in TODO_INSERT_URL before installing v2!</li>
```
Comment (joewiz) [inline repo.xml:17]: Note TODO - I plan to link to the module's GitHub URL - which strikes me as the most direct route to finding it for someone who hasn't yet installed v2.

File: controller.xql (hunk 1 above)
Comment (line-o) [reply controller.xql:100]: No we do not. Well at least it would have been only by accident if, not officially.

Comment (duncdrum) [reply controller.xql:100]: `icon.svg` is supported not sure why you think it isn’t

Comment (joewiz) [reply controller.xql:100]: @duncdrum The code here assumes png and svg. The only question is whether we need to serve others like jpg and gif.

Comment (duncdrum) [reply controller.xql:100]: Ah I misinterpreted your comment, svg and png are the only supported formats.

Comment (joewiz) [reply controller.xql:100]: Got it. Thanks!

File: repo.xml (hunk 2 above)
Comment (joewiz) [reply repo.xml:17]: Alternatively, I could remove the script from the repository and post it as a gist.

**Comment by joewiz** [comment]
Perhaps we can relieve ourselves of the problematic dependency on the crypto lib—by switching to `util:hash()`, which appears to support sha256. The only issue is that it seems to return different values than the `crypto:hash()` used in `modules/scan.xql` with equivalent parameters - same binary, same sha256 algorithm, and same base64 encoding option:

```xquery
xquery version "3.1";

import module namespace crypto = "http://expath.org/ns/crypto";

let $binary := util:binary-doc("/db/apps/public-repo/public/eXide-2.4.10.xar")
return
    (
        util:hash($binary, "sha256", true()),
        crypto:hash($binary, "sha256", "base64")
    )
```

In http://exist-db.org/exist/apps/eXide/index.html (eXist 5.1.1 with This returns:

- util:hash: `cGDaSOudPkk/oV2mumY4QN2/pJq0ou1wP2mje4Q0Zjw=`
- crypto:hash: `3OP7Zo2bgsclVU/y87CZXnq3aV1o/fZ1K9dj6yWKJic=`

Docs:

- util:hash: https://exist-db.org/exist/apps/fundocs/view.html?uri=http://exist-db.org/xquery/util#hash.2
- crypto:hash: http://expath.org/spec/crypto/editor#hash-function

**Comment by joewiz** [comment]
Unfortunately, `util:hash` is just not returning the expected SHA256 values. For example, if you download https://exist-db.org/exist/apps/public-repo/public/eXide-2.4.10.xar:

```bash
$ shasum -a 256 /Users/joe/Downloads/eXide-2.4.10.xar
dce3fb668d9b82c725554ff2f3b0995e7ab7695d68fdf6752bd763eb258a2627
```

The crypto lib returns the same value, but util:hash returns some other value:

```xquery
xquery version "3.1";

import module namespace crypto = "http://expath.org/ns/crypto";

let $value := file:read-binary("/Users/joe/Downloads/eXide-2.4.10.xar")
let $alg := "SHA-256"
return
    (
        crypto:hash($value, $alg, "hex"),
        util:hash($value, $alg)
    )
```

returns:

```text
"dce3fb668d9b82c725554ff2f3b0995e7ab7695d68fdf6752bd763eb258a2627"
"7060da48eb9d3e493fa15da6ba663840ddbfa49ab4a2ed703f69a37b8434663c"
```

@dizzzz As [the creator of util:hash](https://github.com/eXist-db/exist/commit/90d9ac6d5165339ef2a3bfef123e6f7870d8ec2d), do you by chance have any ideas on why I'm not able to get expected results with util:hash when using binaries? The values in the included tests line up with the values I see at https://hashtoolkit.com/generate-hash/?text=ABCDEF, but we don't have any tests for binaries, so I wonder if this area has some problems? 

I ask because eXist's crypto lib isn't compatible with eXist 5.3.0-SNAPSHOT, and crypto:hash is the only reason public repo has a dependency on the crypto lib. If we could use util:hash instead, we could drop our dependency on the crypto lib here. Thanks for any ideas you might have!

**Comment by joewiz** [comment]
@adamretter As the creator of `util:binary-doc-content-digest()`, do you by chance have any ideas why the function is returning an empty sequence? 

```xquery
xquery version "3.1";

let $binary-resource := "/db/eXide-2.4.10.xar"
let $algorithm := "SHA-256"
return
    util:binary-doc-content-digest($binary-resource, $algorithm)
```

If you store https://exist-db.org/exist/apps/public-repo/public/eXide-2.4.10.xar as `/db/eXide-2.4.10.xar`, this query should return:

```text
dce3fb668d9b82c725554ff2f3b0995e7ab7695d68fdf6752bd763eb258a2627
```

Instead it's returning an empty sequence.

**Comment by adamretter** [comment]
@joewiz Hmm... I am not getting the same results as you on 5.3.0-SNAPSHOT.

How are you running this? Is it possible that you are running this in eXide? I ask because in the past there have been experiences of eXide giving some un-true/missing representations of results?

If you run this in the Java Admin Client you see the correct Hex encoded SHA-256 digest:
Screenshot 2021-02-15 at 09 50 52

However, if you run this in the Java Admin Client:
Screenshot 2021-02-15 at 09 52 37

You see the raw SHA-256 (without Hex encoding). This is due to the fact that the Java Admin Client uses the XML:DB API, and that at the moment exposes the raw binary values behind an `xs:base64Binary` or `xs:hexBinary`. Whilst the result is technically correct, for display purposes the Java Admin Client should be improved to show the encoded version. I consider that likely to be a separate bug/enhancement compared to what you have reported though.

File: controller.xql [hunk 3]
```diff
@@ -75,37 +76,47 @@ else if (ends-with($exist:resource, ".html") and starts-with($exist:path, "/pack
...
+    <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
+        <forward url="../../public-repo-data/icons/{$exist:resource}"/>
```
Comment (line-o) [inline controller.xql:102]: Can we still have a variable collection where XARs are stored instead of a hard-coded path?

**Review by line-o — COMMENTED** [review]
Thanks for all the hard work you are putting into this. 💯 
A few questions and details remain.
1. hard coded collection where XARs are stored
2. Should scanrepo.xq be renamed to scanrepo.xqm?
3. Git likes newlines at the end of each file

File: pre-install.xq [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+(: path to the directory containing the unpacked .xar package :)
+declare variable $dir external;
+(: the target collection into which the app is deployed :)
+declare variable $target external;
+
+(: create public-repo-data and subcollections :)
+xmldb:create-collection("/db/apps", "public-repo-data"),
+for $col in ("icons", "metadata", "packages")
+return
+    xmldb:create-collection("/db/apps/public-repo-data", $col)
```
Comment (line-o) [inline pre-install.xq:18]: Here is the missing newline

**Review by line-o — COMMENTED** [review]
I also noticed that the wording changed from "app" to "package(-group)". A change that I support. Would it be possible to rename the functions like "list-apps" to "list-packges" too?

File: controller.xql (hunk 3 above)
Comment (joewiz) [reply controller.xql]: @line-o I've now removed hardcoded collection names here and throughout the app. However, in this one instance in controller.xql, I haven't tried to resolve the arbitrary absolute path defined in config.xqm to the relative path needed in the forward directive.

File: pre-install.xq (hunk 4 above)
Comment (joewiz) [reply pre-install.xq]: @line-o Ok, I've fixed this and the other instances of this in the repository.

**Comment by joewiz** [comment]
@line-o I've now applied the new wording across the app. Here's a fuller description of what I did, taken from the last commit's commit message:

> Toward tackling technical debt
> 
> - instead of making only minimal changes to achieve the goal of the PR, this commit comprehensively updates the app to reflect the goal of the PR
> - apply newly adopted wording for package metadata structures to all functions, so what was “app” is now “package”, “apps” is “package-group”, etc. applied the new naming consistently.
> - use config.xqm to centrally define the names of collections for metadata storage, as well as to store values like the default eXist version (2.2.0) and processor URI; similarly, read the repo user and group centrally from repo.xml instead of sprinkling hardcoded references to these values where needed.
> - refactor the publish/upload routing by only updating the package-group of the uploaded apps, instead of regenerating all package groups
> - only calculate a xar package’s sha-256 hash once instead of twice/repeatedly
> - restructure package-groups to not store any package-specific information at the top level; instead, only package identifiers (name, abbrev, title) are stored there, and all package-specific information is grouped in a child packages element
> - fix the problem where a listing of packages compatible with an older version of eXist shows incorrectly package information for the latest (incompatible) version. now all views correctly limit themselves to compatible versions. any incompatible versions that must be shown are
> - eliminate redundant functions for finding compatible packages by creating a new versions.xqm library module
> - add comments to every module header and to many functions
> - add namespaces and module import declarations for all referenced libraries in every module

**Comment by joewiz** [comment]
@adamretter Also in my last commits, I removed the crypto lib, replacing crypto:hash with util:binary-doc-content-digest (see https://github.com/eXist-db/public-repo/pull/53/commits/5afa0b49c355990baadd1b4d5a85d15248c0f93f). Then I trimmed it down from 2 calls to one (see https://github.com/eXist-db/public-repo/blob/90eb6bb9f50df93ba6473a2d866fb2ae2f89d7b8/modules/scan.xqm#L182). Thanks for pointing me in the right direction!

File: controller.xql (hunk 3 above)
Comment (joewiz) [reply controller.xql]: @line-o Also, I'd already changed scanrepo's file extension from xql to xqm.

**Review by line-o — COMMENTED** [review]
A LOT to take in. From looking at the code it is getting better and better. It is starting to look like a rewrite :)
I will test this

**Comment by joewiz** [comment]
@line-o Thanks for looking! Yes, I agree that it's approaching a rewrite. Sorry I couldn't make the commits more atomic, but there were too many interrelated changes. After many years of making minimal changes just to keep the thing working, we had accumulated some technical debt. So this commit is really a detangling effort, and I hope it facilitates reworking the app further if we decide to tackle some improvements. 

For example, great next steps would be to add some CI tests to prevent regressions and to roasterize the endpoints to formalize the API. 

Stepping back from public-repo alone, though, this app is also part of an ecosystem—stretching from the package spec itself to packageservice, Dashboard's package manager, eXist's internal repo deployment, and exist-generator— that I fear may have diverged in some ways. For example, the package spec was developed before semantic versioning 2.0, a system we have adopted in large part, but some of the older version expressions seem to be (?) allowed and are in the documentation. The package spec also lacks expressiveness that npm enjoys; for example we can say semver-min=2.0.0 but we have no way to say "up to but not including 2.0.0 or pre-release builds," so I've resorted to kludges like semver-max=1.99.99. Similarly, the documentation issue in which we uncovered [additional installation actions available to repo.xml](https://github.com/eXist-db/documentation/issues/164) that aren't really documented or widely used is one example. There might be capabilities that people could find useful. 

A useful task, then, might be to survey how the various package-related clients and utilities implement their package-related operations. Then we could understand the implications of updating the package spec to allow more expressive dependency formulations, or delegating some layers of package management to some standard services. It would be nice not to reinvent the wheel.

Then there's the problem of the many legacy packages published to the current public repo. There are some real problems, like packages that are not compatible with new versions of eXist that lack eXist version declarations. But eXist version dependencies shouldn't be required for all packages, since some library packages like FunctX are standard XQuery and should always work with eXist. It's also not just old packages; as the crypto library shows, it's far too easy for developers to muck up their metadata. We need some package validation tool like epubcheck to flag problems and let us preview how a package's metadata will be interpreted by the various components. We could also use some editorial tools - so that we can mark submissions as having been tested against various versions of eXist and open them up for use with new versions or override the supplied metadata. 

It might not even be practical to review all old apps—but better to start a new repo that has better controls from the start, and a plan for handling compatibility with future versions of eXist. For example, we might want to say that pure XQuery apps should be evergreen, but apps with jars should be limited to a single major version of eXist, to prevent it from automatically getting picked up and installed into new versions.

(I also put some thoughts along these lines, so I won't repeat them here... https://gist.github.com/joewiz/c2b8cdabb80b03234efa915f47df13f9.)

File: controller.xql [hunk 5]
```diff
@@ -53,10 +54,10 @@ else if ($exist:resource eq "update.xql") then
 else if ($exist:path eq "/admin.html") then
     let $user := request:get-attribute("org.exist.public-repo.login.user")
     return
-        if ($user and (sm:is-dba($user) or "repo" = sm:get-user-groups($user))) then
+        if (exists($user) and sm:get-user-groups($user) = "repo") then
```
Comment (joewiz) [inline controller.xql:57]: I should probably update this to use the technique used elsewhere in the PR: Replace the hardcoded `"repo"` with `config:repo-descriptor()/repo:permissions/@group`.

**Comment by joewiz** [comment]
@line-o Thanks for your offer to test! I just updated the list of tests in the main post above, fleshing out which pages and endpoints are there.

File: repo.xml (hunk 2 above)
Comment (joewiz) [reply repo.xml]: I've already updated the link to point to the module's GitHub URL.

**Comment by line-o** [comment]
Testing find.xql status quo:
exist 5.2.0 / public-repo 1.0.3
http://localhost:8080/exist/apps/public-repo/find?abbrev=xbow&processor=5.2.0
gets redirected to 
http://localhost/exist/apps/public-repo/public/xbow-2.0.0.xar 
instead of 
http://localhost:8080/exist/apps/public-repo/public/xbow-2.0.0.xar 
is that a known issue?

**Comment by joewiz** [comment]
@line-o I've committed a fix to the issue you reported with the port number being stripped from the redirect. It works for me. If merged, it'll also close https://github.com/eXist-db/public-repo/issues/49.

File: controller.xql (hunk 5 above)
Comment (joewiz) [reply controller.xql]: Fixed in https://github.com/eXist-db/public-repo/pull/53/commits/166cd0bea1b06d7da55146e8980217ed40106581.

File: controller.xql (hunk 3 above)
Comment (joewiz) [reply controller.xql]: @line-o Ok, I've fixed this problem of the hard-coded relative path in controller.xql. See https://github.com/eXist-db/public-repo/pull/53/commits/b65a476d16148e5bc7bc6b13bf7d6b70124a5385.

File: controller.xql (hunk 5 above)
Comment (joewiz) [reply controller.xql]: To make this information easier to access, I've added a function, `config:repo-premissions()`, which returns a map, so you can simply append `?user` or `?group` to access these values.

**Comment by joewiz** [comment]
@line-o I've just completed work on another round of fixes addressing your questions and fixing some other issues I spotted along the way. This time the commits are more atomic and thus hopefully more easily digested ;)

**Comment by joewiz** [comment]
I forgot to mention that I also took the opportunity in https://github.com/eXist-db/public-repo/pull/53/commits/90eb6bb9f50df93ba6473a2d866fb2ae2f89d7b8#diff-1aadcf801882031bdd8965cd9217d91830c286bc19af70210e00d789e7965432 to switch from [eXist's implementation-specific JSON serializer](https://exist-db.org/exist/apps/wiki/blogs/eXist/JSONSerializer) to XQuery 3.1. I'd like to apply that change to the other stock apps too.

**Comment by line-o** [comment]
To summarize: the Full Monty

**Review by line-o — APPROVED** [review]
A fabulous next version of the public package repository!

Thanks again @joewiz for putting in the time.

PS: I found an edge case and took the liberty to fix it an modify your PR directly, I hope this is OK with you.

**Comment by line-o** [comment]
All tests were successful, the package repository was re-installed several times without any data-loss.

**Comment by joewiz** [comment]
@line-o Excellent! Thanks for the good catch and all of the testing!
